Add API to register callback to set socket options

nosnilmot opened the pull request:
The registered callback function will be called before connect() is called on the socket, allowing tweaking of much more than just keepalive settings.

This deprecates xmpp_conn_set_keepalive(), which will be ignored if it is called for a conn object that already has a sockopt callback set.

Includes helper/example callback function to set default keepalive parameters that some library users may find useful.

This comes from discussion on PR #199.

Examples have been modified to leverage new API:
- bot.c uses included helper callback to set default keepalive values with new --tcp-keepalive option (also reworked option parsing to match basic.c and complex.c)
- complex.c uses custom callback function (but essentially does the same thing)

sock.c is no longer completely standalone from the rest of strophe, as it needed to know about xmpp_conn_t and xmpp_sockopt_callback
We could either:
- live with it? I'm not sure if sock.[ch] are intended to be reusable elsewhere without modification
- move call to callback into conn.c/_conn_connect (I did this originally, but it means callback cannot set socket options before connect() call)
- something else?

sjaeckel commented on src/conn.c:
```diff
@@ -245,6 +262,56 @@ void xmpp_conn_set_keepalive(xmpp_conn_t *conn, int timeout, int interval)
+ *
+ *  @ingroup Connections
+ */
+int xmpp_sockopt_cb_keepalive(xmpp_conn_t *conn, void *socket)
```
I very much like! :+1: 

Can you please move this to `sock.c`? like that you can also remove the `ifdef` and use `sock_t sock = *((sock_t *)socket);` instead

nosnilmot replied:
I didn't do this originally because `sock.c` was independent from rest of libstrophe, but since I then changed that to call the callback function from sock.c anyway ...

sjaeckel commented on src/conn.c:
```diff
@@ -245,6 +262,56 @@ void xmpp_conn_set_keepalive(xmpp_conn_t *conn, int timeout, int interval)
+    int sock = *((int *)socket);
+#endif
+
+    (void)conn;
```
```suggestion
    UNUSED(conn);
```

nosnilmot replied:
Ah yes, I forgot to go back and correct these after I realised the UNUSED macro existed.

sjaeckel commented on src/sock.c:
```diff
@@ -109,12 +116,18 @@ sock_t sock_connect(const char *host, unsigned short port)
     int optval = (timeout && interval) ? 1 : 0;
 
-    /* This function doesn't change maximum number of keepalive probes */
+#ifndef TCP_KEEPCNT
+    (void)count;
+#endif
+#ifndef TCP_USER_TIMEOUT
+    (void)user_timeout;
+#endif
```
```suggestion
    UNUSED(count);
    UNUSED(user_timeout);
```

Removing the `ifndef` shouldn't hurt or not?

If you think it has its advantage to only mark them as `UNUSED` when the value isn't defined, please put the `UNUSED()` in the according `#else` path.

sjaeckel commented on src/sock.c:
```diff
@@ -149,6 +162,30 @@ int sock_set_keepalive(sock_t sock, int timeout, int interval)
+                         sizeof(user_timeout));
+        if (ret < 0)
+            return ret;
+#elif defined(TCP_RXT_CONNDROPTIME)
+        int rxt = user_timeout / 1000;
+        ret = setsockopt(sock, IPPROTO_TCP, TCP_RXT_CONNDROPTIME, &rxt,
+                         sizeof(rxt));
+        if (ret < 0)
+            return ret;
```
:+1:

sjaeckel commented on examples/basic.c:
```diff
@@ -15,10 +15,6 @@
 
 #include <strophe.h>
 
-/* hardcoded TCP keepalive timeout and interval */
```
:+1: 

thanks for cleaning up!

sjaeckel commented on src/conn.c:
```diff
@@ -245,6 +262,56 @@ void xmpp_conn_set_keepalive(xmpp_conn_t *conn, int timeout, int interval)
+void xmpp_conn_set_sockopt_callback(xmpp_conn_t *conn,
+                                    xmpp_sockopt_callback callback)
+{
+    strophe_debug(conn->ctx, "xmpp", "Registering sockopt callback");
+
```
```suggestion
```

sjaeckel reviewed (changes requested):
Thanks! Looks very good.

nosnilmot commented:
Do you prefer I squash commits & force push, or will you squash/rebase during merge?

sjaeckel commented:
> Do you prefer I squash commits & force push, or will you squash/rebase during merge?

Please always squash/fixup, rebase and force-push!

nosnilmot commented:
> Can you please check whether I missed something in [163e34d](https://github.com/strophe/libstrophe/commit/163e34d9340b61c7af75d26a96fa94b87cb79076) or if that's OK and still does everything we want and discussed?

hmm, this introduces some subtle behaviour changes:
- TCP_KEEPCNT will be set for any existing users of xmpp_conn_set_keepalive() (this was the case before 5518a9e10add32608a347103e235597da52779ce too, there was presumably _some_ reason it was removed).
- TCP_USER_TIMEOUT will be set for existing users of xmpp_conn_set_keepalive()
- It's no longer possible to differentiate between xmpp_conn_set_keepalive() / xmpp_conn_set_sockopt_callback() and warn about deprecation of xmpp_conn_set_keepalive(). Maybe this isn't critical.

I think the existing behaviour is restored with 0001773a62d6b73ce896b904691c8a06803117bc.

> `* This function is only effective on a disconnected object.`

That's a good point, and slight loss of functionality. Should we call the callback immediately for a connected object? or change xmpp_conn_set_sockopt_callback() to return the socket (or NULL if not connected) so the user can choose? 

> You could also squash [b798e90](https://github.com/strophe/libstrophe/commit/b798e90bd3c416541c842c005e6c9e9875fde2e1) into [af22d87](https://github.com/strophe/libstrophe/commit/af22d87258abb36396a1716856a2a7a90362f9f6) if you want to.

Will wait to hear if you want to keep 0001773a62d6b73ce896b904691c8a06803117bc or not before cleaning up / rewriting git history.

sjaeckel commented:
> I think the existing behaviour is restored with [0001773](https://github.com/strophe/libstrophe/commit/0001773a62d6b73ce896b904691c8a06803117bc).

Thanks! Please keep it, squash stuff together and rebase.

> > `* This function is only effective on a disconnected object.`
> 
> That's a good point, and slight loss of functionality. Should we call the callback immediately for a connected object? or change xmpp_conn_set_sockopt_callback() to return the socket (or NULL if not connected) so the user can choose?

We're not sure whether libstrophe will only have a single socket per connection in the future (some comment from @pasis in the past), so returning the socket is no option.
I don't see a reason why we shouldn't call the callback immediately for a connected object. IIRC some of the socket options can only be applied to disconnected sockets, if that's the case we should write down in our documentation that the user has to take care of that!

nosnilmot commented:
> > I think the existing behaviour is restored with [0001773](https://github.com/strophe/libstrophe/commit/0001773a62d6b73ce896b904691c8a06803117bc).
> 
> Thanks! Please keep it, squash stuff together and rebase.

done.

> I don't see a reason why we shouldn't call the callback immediately for a connected object. 

added.

> IIRC some of the socket options can only be applied to disconnected sockets, if that's the case we should write down in our documentation that the user has to take care of that!

I'd like to think it was obvious that if you want to set options that can only be set on disconnnected sockets, then the callback should be registered before connection, but I have added a note to that effect anyway.

sjaeckel commented on src/sock.c:
```diff
@@ -149,11 +161,53 @@ int sock_set_keepalive(sock_t sock, int timeout, int interval)
+ *  An example function that can be used to set reasonable default keepalive
+ *  options on sockets when registered for a connection with
+ *  xmpp_conn_set_sockopt_callback()
+ *
```
```suggestion
 *
 *  @param conn a Strophe connection object
 *  @param socket pointer to a socket descriptor
 *
 *  @see xmpp_sockopt_callback for details on the `socket` parameter
```

just for the sake of completeness :)

nosnilmot commented on strophe.h:
````diff
@@ -269,6 +269,43 @@ typedef void (*xmpp_conn_handler)(xmpp_conn_t *conn,
+ *  ```
+ *
+ *  This function will be called for each socket that is created and in case
+ *  this call-back is set, the values set via `xmpp_conn_set_keepalive()` are
+ *  ignored.
````
corrected

nosnilmot commented on src/conn.c:
```diff
@@ -226,23 +254,45 @@ xmpp_conn_t *xmpp_conn_clone(xmpp_conn_t *conn)
  *  @param timeout TCP keepalive timeout in seconds
  *  @param interval TCP keepalive interval in seconds
  *
+ *  @note this function is deprecated
```
Why didn't I see that before? Oh... it's brand new. yes, done that.

sjaeckel reviewed (approved):
another squash and we're fine to go!

thanks a lot :)

nosnilmot commented on examples/complex.c:
```diff
@@ -166,9 +168,9 @@ static int sockopt_cb(xmpp_conn_t *conn, void *socket)
 #endif /* _WIN32 */
-    return ret;
+
+    return 0;
```
was this change intentional?

sjaeckel replied:
All error handling is done immediately after it could occur or not? i.e. we would only arrive here when `ret` is `0`!?

nosnilmot replied:
I reverted it in the final squash because that is what is returning success/fail for the _WIN32 case (same is in sock.c/sock_set_keepalive())

sjaeckel replied:
then it was accidental :smile:
